### ISSUE TITLE
Resolve agent and hostname issues when Amazon/AWS cloud provider is configured

### ISSUE THREAD

**Issue by superseb** [rancher/rancher#22849]
When cloud provider is configured to Amazon/AWS, the node resource name is set by kubelet to the internal DNS name as described in the AWS cloud provider documentation. This means that we create a node with (host)name based on the node template, launch agents using these values but the node resource in Kubernetes gets changed for this node as soon as the kubelet starts and registers itself to the cluster. This creates two issues:

- The `share-mnt` container makes sure that on reboot, the agent is respawned to write certificates to the host. On worker nodes where `/etc/` is non-persistent, this means that that agent is responsible for putting the certificates back before starting the cluster components like kubelet. Because the initial hostname that the `share-mnt` is spawned with, does not match the new name that the node got after it registered and got the name that is set by the cloud provider, it is not allowed to connect and logs `bad handshake`. The actual error is `machine not found`, as the machineLister on the Rancher side tries to find a node with the initial hostname while its changed to the new cloud-provider set hostname. Possible solution here is to have the machineLister also check the `requestedHostname` for the node as this still contains the original node name.

Workarounds for this issue are in https://github.com/rancher/rancher/issues/17431#issue-399093017

- We should investigate what to do with the configured hostname/requestedHostname/labels when the cloud provider is set to AWS and the node resource name is changed. One use case is mentioned in https://github.com/rancher/rancher/issues/22838:

```
The biggest problem I'm running into right now is that when RancherOS gets added to kubernetes, the kubernetes.io/hostname gets set to the friendly-name/requested-name, but the node name in kubernetes is the AWS ip-X-X-X-X hostname. This leads to the inability to schedule OpenEBS volumes for example because it creates scheduling rules based on the node name targeting the kubernetes.io/hostname which do not match.
```

**Comment by ekristen** [comment]
FYI -- https://github.com/kubernetes/cloud-provider-aws/issues/57

**Comment by ibrokethecloud** [comment]
I send this custom user-data via the node template:

<img width="1246" alt="image" src="https://user-images.githubusercontent.com/19664014/65306994-1bee7580-dbca-11e9-9978-0f1c0f8bd7ee.png">

```
#cloud-config
bootcmd:
 - mkdir -p /var/log/kubernetes
 - ln -s /var/log/kubernetes /etc/kubernetes
```

I have tested it via a node template and it seems to work fine.

Could be a possible work around while for now.

**Comment by superseb** [comment]
This needs to be heavily tested for all nodes in a cluster, especially in an upgrade situation.

**Comment by jiaqiluo** [comment]
Te bug is reproduced in Rancher: v2.3.1

Steps:
- add an ec2 cluster using RancherOS and cluster provider enabled
- 1 etcd, 1 control plane, 1 worker
- after the cluster is active, reboot the worker node from AWS console

Results:
- the worker node is in an unavailable state in UI

<img width="1379" alt="Screen Shot 2019-11-05 at 5 05 11 PM" src="https://user-images.githubusercontent.com/6218999/68256950-e58c7e80-ffee-11e9-9c55-30d127a18811.png">

- ssh into the worker node, check the log of the container with image `rancher/rancher-agent:v2.3.1`
```
time="2019-11-06T00:03:32Z" level=error msg="Failed to connect to proxy. Response status: 200 - 200 OK. Response body: node.management.cattle.io \"c-lf448/m-98851688d825\" not found" error="websocket: bad handshake"
time="2019-11-06T00:03:32Z" level=error msg="Remotedialer proxy error" error="websocket: bad handshake"
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in Rancher:v2.3-head 4f04dda20

Steps:
- add an ec2 cluster using RancherOS and cluster provider enabled
- 1 etcd, 1 control plane, 1 worker
- after the cluster is active, reboot the worker node from AWS console

Results:
- the worker node is reconnected and active
- the cluster is still functional

**Comment by jiaqiluo** [comment]
need to do more tests around upgrading

**Comment by jiaqiluo** [comment]
The bug fix is validated with the following upgrade path

- run Rancher:v2.3.2
- add an RKE cluster: `Rancher OS`, `cloud provider `enabled
- download the key for the node from UI
- ssh into to node 
- check the container `share-mnt` 
```
9e5c0d32b393        rancher/rancher-agent:v2.3.2         "run.sh -- share-roo…"   2 hours ago         Exited (0) 2 hours ago                       share-mnt
```
- check its spec, notice the  the `Args` does NOT set `node-name` 
```
"Args": [
            "--",
            "share-root.sh",
            "docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/var/run rancher/rancher-agent:v2.3.2 --server https://159.65.255.104 --token 56jnlsvl9hbhklk9h4p2zgs6wbfqjsqgjrhhg6cq5fl8rtcpgwtlbb --ca-checksum 772983ef1a697e9ed8e97ccaeddbf5237ef0745ebdf719ca40b670d52b242fac --no-register --only-write-certs",
            "/opt/rke/var/lib/kubelet",
            "/var/lib/rancher",
            "/var/lib/kubelet/volumeplugins"
        ],
```

- upgrade Rancher to `v2.3-head`
- wait for the cluster to be `active` again
- check the container `share-mnt`, notice the container got recreated
```
9f83c20a3193        rancher/rancher-agent:v2.3-head      "run.sh -- share-roo…"   2 minutes ago        Exited (0) 2 minutes ago                       share-mnt
```
- check its spec, notice the `node-name` is set for the `Args`
```
"Args": [
            "--",
            "share-root.sh",
            "docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/var/run rancher/rancher-agent:v2.3-head --server https://159.65.255.104 --token 56jnlsvl9hbhklk9h4p2zgs6wbfqjsqgjrhhg6cq5fl8rtcpgwtlbb --ca-checksum 772983ef1a697e9ed8e97ccaeddbf5237ef0745ebdf719ca40b670d52b242fac --no-register --only-write-certs",
            "--node-name",
            "jiaqi-232-1-15-w2",
            "/opt/rke/var/lib/kubelet",
            "/var/lib/rancher",
            "/var/lib/kubelet/volumeplugins"
        ],
```

- reboot this node
- see the node goes unavailable then back available
- check the container `share-mnt` , notice its the same container, but it ran and exit after rebooting 
```
9f83c20a3193        rancher/rancher-agent:v2.3-head      "run.sh -- share-roo…"   5 minutes ago       Exited (0) 19 seconds ago                         share-mnt
```

More Checks:

Also  check the node in other RKE clusters and confirm the new parameter `--node-name` is added to the `share-mnt` container's arguments

**Comment by kspradheep** [comment]
Hi, @jiaqiluo @superseb  I'm seeing an identical issue for importing GKE v1.15.12 cluster size 50 node count to Rancher HA v2.3.6 or v2.3.9 (hosted in AWS).

Network environment:
private link (GCP network --> internal Megaport network ---> AWS private route)
MTU 1450
No proxy setup behind Rancher

Notes:

- I was able to import GKE v1.14.x cluster to the same Rancher and network environment is similar. I tried cleaning up cluster nodes re-importing the cluster but didn't help much.

- Rancher upgrade from v2.3.6 to v2.3.9 didn't help.

cattle-node-agent logs: Seeing repeated logs of below
time="2020-09-01T13:01:35Z" level=info msg="Connecting to proxy" url="wss://rancher.mydomain.com/v3/connect"
time="2020-09-01T13:01:36Z" level=error msg="Failed to connect to proxy. Response status: 200 - 200 OK. Response body: node.management.cattle.io \"c-6rkb7/m-726d1fcb8f06\" not found" error="websocket: bad handshake"
time="2020-09-01T13:01:36Z" level=error msg="Remotedialer proxy error" error="websocket: bad handshake"

Relevant Rancher logs:
2020/09/01 13:06:14 [DEBUG] machineName: returning [m-726d1fcb8f06] for node with RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-dg4y]
2020/09/01 13:06:14 [DEBUG] getMachine: looking up machine [m-726d1fcb8f06] in cluster [c-6rkb7]
2020/09/01 13:06:24 [DEBUG] machineName: returning [m-726d1fcb8f06] for node with RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-dg4y]
2020/09/01 13:06:24 [DEBUG] getMachine: looking up machine [m-726d1fcb8f06] in cluster [c-6rkb7]
2020/09/01 13:06:56 [DEBUG] machineName: returning [m-726d1fcb8f06] for node with RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-dg4y]
2020/09/01 13:06:56 [DEBUG] getMachine: looking up machine [m-726d1fcb8f06] in cluster [c-6rkb7]
2020/09/01 13:07:17 [DEBUG] machineName: returning [m-726d1fcb8f06] for node with RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-dg4y]
2020/09/01 13:07:17 [DEBUG] getMachine: looking up machine [m-726d1fcb8f06] in cluster [c-6rkb7]
2020/09/01 13:07:39 [DEBUG] machineName: returning [m-726d1fcb8f06] for node with RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-dg4y]
2020/09/01 13:07:39 [DEBUG] getMachine: looking up machine [m-726d1fcb8f06] in cluster [c-6rkb7]

other error Rancher logging:
2020/09/01 13:09:32 [DEBUG] c-6rkb7(f3d776c9-7322-409e-92da-f9a68039d12f): (437058416 * 3) / 4294967295 = [10.42.0.12 10.42.0.13 10.42.0.14][0] = 10.42.0.12, self = 10.42.0.13
2020/09/01 13:09:32 [DEBUG] getMachine: looking up machine [m-d8a679c3a7fd] in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: looking up [gke-los-angeles-main-los-angeles-zale-e81bae54-vl6s] as node name in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: looking up inNode.RequestedHostname [gke-los-angeles-main-los-angeles-zale-e81bae54-vl6s] in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: unable to match inNode.RequestedHostname [gke-los-angeles-main-los-angeles-zale-e81bae54-vl6s] to a node in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: looking up machine [m-8699a5e356a6] in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: looking up [gke-los-angeles-main-los-angeles-defa-7b368758-6qmp] as node name in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: looking up inNode.RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-6qmp] in cluster [c-6rkb7]
2020/09/01 13:09:32 [DEBUG] getMachine: unable to match inNode.RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-6qmp] to a node in cluster [c-6rkb7]
2020/09/01 13:09:33 [DEBUG] getMachine: looking up machine [m-6da69cadb62e] in cluster [c-6rkb7]
2020/09/01 13:09:33 [DEBUG] getMachine: looking up [gke-los-angeles-main-los-angeles-defa-7b368758-xx32] as node name in cluster [c-6rkb7]
2020/09/01 13:09:33 [DEBUG] getMachine: looking up inNode.RequestedHostname [gke-los-angeles-main-los-angeles-defa-7b368758-xx32] in cluster [c-6rkb7]

Environment:
Rancher HA v2.3.9
Ubuntu OS
RKE CLI v1.0.12
Kubernetes V1.17.9
Docker info:
Client:
 Debug Mode: false

Server:
 Containers: 30
  Running: 21
  Paused: 0
  Stopped: 9
 Images: 27
 Server Version: 19.03.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.3.0-1033-aws
 Operating System: Ubuntu 18.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 30.58GiB
 Name: PREORAINFRKE03
 ID: ZTFE:JGGC:3CCH:GDAU:AR33:5Y4T:TGRM:3AKK:VU7A:MWIC:6PUE:J6AS
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support

**Comment by kspradheep** [comment]
It's resolved for me by doing upgrade from v2.3.9 to v2.4.6. However, rancher cli v2.4.6 is not working when trying to add AD groups to cluster (not seeing this issue in UI).

rancher clusters add-member-role <ad group name> cluster-owner

FATA[0005] Bad response statusCode [422]. Status [422 Unprocessable Entity]. Body: [baseType=error, code=InvalidBodyContent, message=must target a user [userId]/[userPrincipalId] OR a group [groupId]/[groupPrincipalId]] from [https://rancher.mydomain.com/v3/clusterroletemplatebindings]

The CLI issue is tracked here: #28718
